runners/gearpump/src/main/java/org/apache/beam/runners/gearpump/translators/GroupByKeyTranslator.java

```diff
@@ -67,66 +67,62 @@ public void translate(GroupByKey<K, V> transform, TranslationContext context) {
+        .reduce(new Merge<>(windowFn, outputTimeFn), "merge")
         .map(new Values<K, V>(), "values");
 
     context.setOutputStream(context.getOutput(transform), outputStream);
   }
 
-  private static class GearpumpWindowFn<T, W extends BoundedWindow> implements WindowFn,
-      Serializable {
+  private static class GearpumpWindowFn<T, W extends BoundedWindow>
+      implements WindowFunction<WindowedValue<T>>, Serializable {
 
-    private org.apache.beam.sdk.transforms.windowing.WindowFn<T, W> windowFn;
+    private final boolean isNonMerging;
 
-    GearpumpWindowFn(org.apache.beam.sdk.transforms.windowing.WindowFn<T, W> windowFn) {
-      this.windowFn = windowFn;
+    public GearpumpWindowFn(boolean isNonMerging) {
+      this.isNonMerging = isNonMerging;
     }
 
     @Override
-    public scala.collection.immutable.List<Bucket> apply(final Instant timestamp) {
+    public Window[] apply(Context<WindowedValue<T>> context) {
       try {
-        Collection<W> windows = windowFn.assignWindows(windowFn.new AssignContext() {
-          @Override
-          public T element() {
-            throw new UnsupportedOperationException();
-          }
-
-          @Override
-          public org.joda.time.Instant timestamp() {
-            return TranslatorUtils.java8TimeToJodaTime(timestamp);
-          }
-
-          @Override
-          public W window() {
-            throw new UnsupportedOperationException();
-          }
-        });
-
-        List<Bucket> buckets = new LinkedList<>();
-        for (BoundedWindow window : windows) {
-          buckets.add(getBucket(window));
-        }
-        return JavaConversions.asScalaBuffer(buckets).toList();
+        return toGearpumpWindows(context.element().getWindows().toArray(new BoundedWindow[0]));
       } catch (Exception e) {
         throw new RuntimeException(e);
       }
     }
 
-    private Bucket getBucket(BoundedWindow window) {
+    @Override
+    public boolean isNonMerging() {
+      return isNonMerging;
+    }
+
+    private Window[] toGearpumpWindows(BoundedWindow[] windows) {
+      Window[] gwins = new Window[windows.length];
+      for (int i = 0; i < windows.length; i++) {
+        gwins[i] = toGearpumpWindow(windows[i]);
+      }
+      return gwins;
+    }
+
+    private Window toGearpumpWindow(BoundedWindow window) {
       if (window instanceof IntervalWindow) {
         IntervalWindow intervalWindow = (IntervalWindow) window;
         Instant start = TranslatorUtils.jodaTimeToJava8Time(intervalWindow.start());
         Instant end = TranslatorUtils.jodaTimeToJava8Time(intervalWindow.end());
-        return new Bucket(start, end);
+        return new Window(start, end);
       } else if (window instanceof GlobalWindow) {
         Instant end = TranslatorUtils.jodaTimeToJava8Time(window.maxTimestamp());
-        return new Bucket(Instant.MIN, end);
+        return new Window(Instant.MIN, end);
```

Comment:
An `IntervalWindow` in Beam represents the interval [`start`, `end`) but actually that is [`start`, `maxTimestamp`] at the resolution of milliseconds. The `maxTimestamp` is 1 millisecond below the `end`. So either the treatment of `IntervalWindow` needs a `.minus(1)` or this treatment of the global window needs a `.plus(1)`.

It would be nice to come up with a test that could catch this off-by-one.